doc/design/fork.md

```diff
@@ -0,0 +1,42 @@
+
+The process clone itself is handled within the kernel, but the fork implementation is actually implemented in an overridden CRT fork implementation. The stack duplication is an example of what is done in user space, versus the clone which happens in the kernel.
+
+ With the single address space being shared between the parent and all forked children (and forks of forked children) there are a number of issues that can arise.
+ The single address space causes the same global variables in the parent to be shared with the children, meaning changes made by one of the processes will affect the other.
+ The C runtime itself has many global variables, and these global variables (along with any application specific ones too) when changed may unexpectedly affect all the other applications sharing the same address space.
+ One important one to note is that of the C runtime itself which would deinitialize the whole CRT on the first process exit causing a cascade of problems and crashes for all other processes.
+ For this specific reason the deinitialization of the CRT needs to be deferred until the last process is exited. For this reason when a child forked process process exits the CRT cleanup code is skipped.
+ Only the parent of all the forked processes will do a CRT deinitialization when exit() is called.
+ This can cause cleanup code that has been queued up for atexit to not run when the application is exited, and may be run (with unspecified behaviour) when the main process exits.
+ To make sure the CRT is only cleaned up by the first process in a fork chain of processes this process needs to either kill all child processes on exit, or wait for them all to finish.
+ This is achieved with the fork implementation chosen.
+
+ The shared address space also causes problems for the copying of the parents stack into the child.
+ The new child stack has all the frame pointers updated such that a stack walk can be achieved in the current stack, any stack variables that may point to other stack variables on the same stack will be incorrect and will point to the parents stack. One  way to achieve this is to search the space on the stack where function parameters and stack variables reside and fix them up, but this is potentially error prone and may cause fixup of other random values on the stack. Therefore these stack pointers are not currently being fixed up.
+
+ Another problem with shared address space is related to the return of the fork() command. The parent may allocate some memory that will be used in the child forked process, but when fork() returns to continue execution the parent may delete this memory, causing the child forked process to use memory after it is freed. For this reason the safest model may be to only support a strict fork/exec model where the parent waits for the child to execute a new process through one of the exec*() APIs. This mode of operation is enabled through the fork configuration.
```

Comment:
Suggested additions:

The scenario that we envision pseudo fork would be useful is:

* Right after the `fork()`, the parent waits for the child to complete either voluntarily (e.g., `waitpid`) or  involuntarily, i.e., `ForkMode = pseudo_wait_for_children`; and
* The child invokes an exec* (execv, execl, etc.) function after the fork; and
* Between the `fork` and the `exec*`, the child only reads from global variables (include CRT global states). It can only write to a stack variable when it is not accessed via a pointer.

We are researching on mechanisms to analyze applications to determine whether such conditions are meet.